tmol/score/cartbonded/potentials/cartbonded_pose_score.impl.hh

```diff
@@ -0,0 +1,392 @@
+
+// Operator definitions; safe for CPU compilation
+#include <moderngpu/operators.hxx>
+
+#include <chrono>
+
+#include "potentials.hh"
+
+namespace tmol {
+namespace score {
+namespace cartbonded {
+namespace potentials {
+
+template <typename Real, int N>
+using Vec = Eigen::Matrix<Real, N, 1>;
+template <typename Real>
+using CoordQuad = Eigen::Matrix<Real, 4, 3>;
+#define Real3 Vec<Real, 3>
+
+enum class subgraph_type { length, angle, torsion };
+
+template <typename Int>
+TMOL_DEVICE_FUNC subgraph_type get_subgraph_type(Vec<Int, 4> subgraph) {
+  if (subgraph[2] == -1 && subgraph[3] == -1) return subgraph_type::length;
+  if (subgraph[3] == -1) return subgraph_type::angle;
+  return subgraph_type::torsion;
+}
+
+template <typename Int, Int size>
+TMOL_DEVICE_FUNC Vec<Int, size> atom_local_to_global_indices(
+    Vec<Int, size> local_indices, Int offset) {
+  Vec<Int, size> global_indices;
+  for (int i = 0; i < size; i++) {
+    if (local_indices[i] != -1)
+      global_indices[i] = local_indices[i] + offset;
+    else
+      global_indices[i] = -1;
+  }
+  return global_indices;
+}
+
+// Get the atom ID for an atom index from a lookup table, preserving (-1)s
+template <typename Int, tmol::Device D>
+TMOL_DEVICE_FUNC Int
+get_atom_id(TensorAccessor<Int, 1, D> atom_id_table, Int atom_index) {
+  return (atom_index == -1) ? -1 : atom_id_table[atom_index];
+}
+
+// From a Vec of atom indices and a lookup table, return a new Vec with their
+// IDs preserving (-1)s
+template <typename Int, Int size, tmol::Device D>
+TMOL_DEVICE_FUNC Vec<Int, size> get_atom_ids(
+    TensorAccessor<Int, 1, D> atom_id_table, Vec<Int, size> atoms) {
+  Vec<Int, size> atom_ids;
+  for (int i = 0; i < size; i++) {
+    atom_ids[i] = get_atom_id(atom_id_table, atoms[i]);
+  }
+  return atom_ids;
+}
+
+// Reverse the non-(-1) elements of a subgraph
+template <typename Int>
+TMOL_DEVICE_FUNC void reverse_subgraph(Vec<Int, 4>& subgraph) {
+  subgraph_type type = get_subgraph_type(subgraph);
+  Int temp;
+  switch (type) {
+    case subgraph_type::length:
+      temp = subgraph[0];
+      subgraph[0] = subgraph[1];
+      subgraph[1] = temp;
+      break;
+    case subgraph_type::angle:
+      temp = subgraph[0];
+      subgraph[0] = subgraph[2];
+      subgraph[2] = temp;
+      break;
+    case subgraph_type::torsion:
+      temp = subgraph[0];
+      subgraph[0] = subgraph[3];
+      subgraph[3] = temp;
+      temp = subgraph[1];
+      subgraph[1] = subgraph[2];
+      subgraph[2] = temp;
+      break;
+  }
+}
+
+template <typename Real, typename Int, int N, tmol::Device D>
+TMOL_DEVICE_FUNC void accumulate_result(
+    common::tuple<Real, Vec<Real3, N>> to_add,
+    Vec<Int, N> atoms,
+    Real& V,
+    TensorAccessor<Vec<Real, 3>, 1, D> dV) {
+  accumulate<D, Real>::add(V, common::get<0>(to_add));
+  for (int i = 0; i < N; i++) {
+    accumulate<D, Vec<Real, 3>>::add(dV[atoms[i]], common::get<1>(to_add)[i]);
+  }
+}
```

Comment:
Here's an example of calling non-lambdas. Still suffers from not being able to capture parent scope, but maybe that would be less of an issue with packing args/vars into structs?